### PR TITLE
CLI: detect and store weights as float16

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Updates the `pt-to-tf` CLI to detect and store weights as float16. Battle-tested with OPT.

### CODE REVIEW

File: src/transformers/commands/pt_to_tf.py [hunk 1]
```diff
@@ -225,8 +226,18 @@ def run(self):
...
         # Save the weights in a TF format (if needed) and confirms that the results are still good
         tf_weights_path = os.path.join(self._local_dir, TF_WEIGHTS_NAME)
+        pt_weights_path = os.path.join(self._local_dir, PT_WEIGHTS_NAME)
         if not os.path.exists(tf_weights_path) or self._new_weights:
             tf_from_pt_model.save_weights(tf_weights_path)
+            # If the resulting TF weights are much bigger than the PT weights (roughly 2x), then we are in the presence
+            # of float16 weights. We can set the corresponding backend, store the weights in the new format, and reset
+            # the backend to ensure the reloaded weights match their counterpart using 32 bit precision.
+            if float(os.path.getsize(tf_weights_path)) > 1.8 * os.path.getsize(pt_weights_path):
+                self._logger.warn("WEIGHT FORMAT UPDATE:float16 weights detected! Storing TF weights in float16")
```
Comment (sgugger) [inline src/transformers/commands/pt_to_tf.py:236]:
```suggestion
                self._logger.warn("WEIGHT FORMAT UPDATE: float16 weights detected! Storing TF weights in float16")
```

**Review by sgugger — COMMENTED** [review]
Will TensorFlow automatically load stored FP16 weights in FP32 like PyTorch does?

File: src/transformers/commands/pt_to_tf.py (hunk 1 above)
Comment (gante) [reply src/transformers/commands/pt_to_tf.py:236]: > Will TensorFlow automatically load stored FP16 weights in FP32 like PyTorch does?

No, if we inspect the variables after loading, they are FP16. However, as soon as we pass some input, we can see that the outputs of the internal computations are FP32, despite the weights being FP16. 

The model error (vs PT) is exactly the same, before and after adding these lines :)

Comment (gante) [reply src/transformers/commands/pt_to_tf.py:236]: The documentation is not clear, but I believe `tf.keras.backend.set_floatx` sets the precision of the internal computations. For instance, if we don't reset to float32 after storing as float16, the error is much much larger (~1e3 times larger).

Comment (gante) [reply src/transformers/commands/pt_to_tf.py]: replied in the wrong place 🤦  moving the discusson

**Comment by gante** [comment]
> Will TensorFlow automatically load stored FP16 weights in FP32 like PyTorch does?

No, if we inspect the variables after loading, they are FP16. However, as soon as we pass some input, we can see that the outputs of the internal computations are FP32, despite the weights being FP16.

The model error (vs PT) is exactly the same, before and after adding these lines :)

The documentation is not clear, but I believe tf.keras.backend.set_floatx sets the precision of the internal computations. For instance, if we don't reset to float32 (the default) after storing as float16, the error is much much larger (~1e3 times larger).

**Comment by sgugger** [comment]
Mmmm, in this case I would avoid storing the weights in FP16 on the Hub before adding something in `from_pretrained` that will convert them back to FP32 like it's done for PyTorch.

**Comment by gante** [comment]
> Mmmm, in this case I would avoid storing the weights in FP16 on the Hub before adding something in from_pretrained that will convert them back to FP32 like it's done for PyTorch.

👍 I can give it a go (and leave this PR open meanwhile)

**Pull request closed without merging**